File: lib/agent/twilio.js
const BbPromise = require('bluebird')
const twilio = require('twilio')
const logger = require('log4js').getLogger('twilio')
const authSessionManager = require('../authSessionManager')
const networkManager = require('../networkManager')
const util = require('../util')

const ROUTER_BASE_PATH = '/twilio'

class TwimlServer {

  constructor (webServer, authToken) {
    this.statusCbBasePath = '/statusCb/failure/'
    this.authActionBasePath = '/actions/auth/'
    const router = webServer.createRouter(ROUTER_BASE_PATH)
    const twilioWebHook = twilio.webhook(authToken)
    router.post(`${this.statusCbBasePath}:sessionId`, twilioWebHook, this.failedCbMiddleware.bind(this))
    router.post(`${this.authActionBasePath}:sessionId`, twilioWebHook, this.authActionMiddleware.bind(this))
    this.sessions = {}
    this.webServer = webServer
  }

  getStatusCbUrl (sessionId) {
    return networkManager.getExternalIp().then(ip => `http://${this.webServer.authUsername}:${this.webServer.authPassword}@${ip}:3000${ROUTER_BASE_PATH}${this.statusCbBasePath}${sessionId}`)
  }

  getAuthActionUrl (sessionId) {
    return networkManager.getExternalIp().then(ip => `http://${this.webServer.authUsername}:${this.webServer.authPassword}@${ip}:3000${ROUTER_BASE_PATH}${this.authActionBasePath}${sessionId}`)
  }

  failedCbMiddleware (req, res) {
    const session = authSessionManager.getSession(req.params.sessionId)
    if (session) {
      session.reportFailure(new Error('Call ended with status ' + req.body.CallStatus), 'twilio')
    }
    res.status(200).end()
  }

  sendTwiml (res, twiml) {
    const rawTwiml = twiml.toString()
    logger.debug('raw twiml ' + rawTwiml)
    res.status(200)
    res.send(rawTwiml)
  }

  authActionMiddleware (req, res) {
    logger.debug('Got req %j', req.body)
    const session = authSessionManager.getSession(req.params.sessionId)
    const twiml = new twilio.TwimlResponse()
    res.set('Content-Type', 'text/xml')
    if (!session) {
      twiml.say('Authentication failed, goodbye !')
      return this.sendTwiml(res, twiml)
    }
    logger.debug('session %s', session)

    if (session.isAuthSuccessful() && session.twilio.nextStep !== 'validateDisarmTime') { // case auth is done by other authenticator
      twiml.say('You have been authenticated, goodbye')
      return this.sendTwiml(res, twiml)
    } else if (session.isAuthAborted()) {
      twiml.say('Authentication is not required anymore')
      return this.sendTwiml(res, twiml)
    } else if (session.isAuthFailed()) {
      twiml.say('Authentication timed out')
      return this.sendTwiml(res, twiml)
    }

    this.getAuthActionUrl(req.params.sessionId)
      .then(actionUrl => {
        if (session.twilio.nextStep === 'askPassword') {
          session.twilio.nextStep = 'verifyPassword'
          twiml.say('Hi this is your rpicalarm speaking')
            .gather({
              action: actionUrl,
              timeout: 30,
              finishOnKey: '#'
            }, function () {
              this.say('Please enter your password followed by the pound key')
            })
        } else if (session.twilio.nextStep === 'validateDisarmTime') {
          try {
            const duration = util.parsePhoneDigitsDuration(req.body.Digits)
            twiml.say(`Thanks, alarm disarmed for ${duration.humanize()}`)
            session.setDisarmDuration(duration, 'twilio')
          } catch (err) {
            twiml.say(`Invalid duration`)
              .gather({
                action: actionUrl,
                timeout: 30,
                finishOnKey: '#'
              }, function () {
                this.say('enter disarm time, last digit is the unit, followed by the pound key')
              })
          }
        } else if (session.twilio.nextStep === 'verifyPassword') {
          if (!session.verify(req.body.Digits)) {
            ++session.twilio.tries
            if (session.maxTries === session.twilio.tries) {
              session.reportAuthFailure('Max tries reached', 'twilio')
              twiml.say('Maximum number of tries reached, goodbye')
            } else {
              const remainingTries = session.maxTries - session.twilio.tries
              twiml.say(`Authentication failed, you have ${remainingTries} ${remainingTries > 1 ? 'tries' : 'try'} remaining`)
                .gather({
                  action: actionUrl,
                  timeout: 30,
                  finishOnKey: '#'
                }, function () {
                  this.say('Please enter your password followed by the pound key')
                })
            }
          } else {
            session.twilio.nextStep = 'validateDisarmTime'
            session.reportAuthSuccess()
            twiml.say(`Authentication succeeded`)
              .gather({
                action: actionUrl,
                timeout: 30,
                finishOnKey: '#'
              }, function () {
                this.say('enter disarm time, last digit is the unit')
              })
          }
        }
        return this.sendTwiml(res, twiml)
      })
      .catch(err => {
        session.reportFailure(err, 'twilio')
        twiml.say(`Authentication succeeded`)
        logger.error('Failure occurred', err)
        this.sendTwiml(res, twiml)
      })
  }
}

class TwilioAgent {

  constructor ({
    account_sid: accountSid,
    auth_token: authToken,
    landline_phone_number: landlinePhoneNumber,
    mobile_phone_number: mobilePhoneNumber
  }, {
    webServer
  }) {
    this.twilioClient = twilio(accountSid, authToken)
    this.landlinePhoneNumber = landlinePhoneNumber
    this.mobilePhoneNumber = mobilePhoneNumber
    this.twimlServer = new TwimlServer(webServer, authToken)
  }

  authenticate (session) {
    logger.debug('Starting twilio authentication')
    if (session.isAuthDone()) {
      return BbPromise.resolve()
    }
    session.twilio = {
      tries: 0,
      nextStep: 'askPassword'
    }
    BbPromise.mapSeries([this.twimlServer.getAuthActionUrl, this.twimlServer.getStatusCbUrl], res => res.call(this.twimlServer, session.id))
      .then(([authActionUrl, statusUrl]) => {
        return new BbPromise((res, rej) => {
          this.twilioClient.calls.create({
            url: authActionUrl,
            to: this.mobilePhoneNumber,
            from: this.landlinePhoneNumber,
            statusCallback: statusUrl,
            statusCallbackEvent: ['busy', 'failed', 'no_answer', 'canceled']
          }, (err, call) => {
            if (err) {
              rej(err)
              return
            }
            logger.info(`Called made to [${this.mobilePhoneNumber}] from [${this.landlinePhoneNumber}] with uri [${call.uri}]`)
          })
        })
      })
      .catch(err => {
        session.reportFailure(err, 'twilio')
      })
  }
  notify (date) {
    // Not working
    // SMS requires to pay a phone number to set the from address
    return new BbPromise((res, rej) => {
      this.twilioClient.messages.create({
        to: this.mobilePhoneNumber,
        from: 'rpicalarm',
        body: '[rpicalarm] intrusion detected at ' + date
      }, (err, message) => {
        if (err) {
          return rej(err)
        }
        return res()
      })
    })
  }
}
TwilioAgent.supports = ['authentication']
TwilioAgent.$inject = ['webServer']

module.exports = TwilioAgent

File: rpicalarm/agents/twilio.py
# -*- coding: utf-8 -*-

from threading import Timer

from flask import abort, request
from twilio.twiml.voice_response import VoiceResponse, Gather
from twilio.request_validator import RequestValidator
from twilio.rest import Client

from .. import events, getLogger, AuthFailureReason, AlarmState, parse_duration

LOGGER = getLogger(__name__)

GENERIC_ERROR = VoiceResponse()
GENERIC_ERROR.say("An error occurred")
GENERIC_ERROR.say("Goodbye")
GENERIC_ERROR.hangup()

DIGITS_TO_UNIT = {
    "4": "h",
    "3": "d",
    "6": "m",
    "7": "s"
}


def twilio_validation_decorate(func, token):
    def twilio_validation_decorated(*args, **kwargs):
        # Create an instance of the RequestValidator class
        validator = RequestValidator(token)

        # Validate the request using its URL, POST data,
        # and X-TWILIO-SIGNATURE header
        request_valid = validator.validate(
            request.url,
            request.form,
            request.headers.get('X-TWILIO-SIGNATURE', ''))

        # Continue processing the request if it's valid, return a 403 error if
        # it's not
        if request_valid:
            LOGGER.debug("Request is valid continuing args=%s,kwargs=%s", str(args), str(kwargs))
            return func(*args, **kwargs)
        else:
            return abort(403)
    return twilio_validation_decorated


class TwilioServer(object):

    def __init__(self, twilio_agent):
        self.alarm = twilio_agent.alarm
        self.web_server = twilio_agent.web_server
        self.twilio_agent = twilio_agent
        self.web_server.add_route(
            "/twilio/auth/<session_id>", "auth", twilio_validation_decorate(self.auth_request, self.twilio_agent.auth_token), methods=["POST"])

        self.web_server.add_route(
            "/twilio/callback/<session_id>", "callback", twilio_validation_decorate(self.callback_request, self.twilio_agent.auth_token), methods=["POST"])

    def get_gather(self, session_id):
        return Gather(action=self.get_auth_action_url(session_id),
                      timeout=30,
                      finishOnKey="#")

    def get_auth_action_url(self, session_id):
        return "{}/twilio/auth/{}".format(self.web_server.auth_base_url, session_id)

    def get_status_callback_url(self, session_id):
        return "{}/twilio/callback/{}".format(self.web_server.auth_base_url, session_id)

    def callback_request(self, session_id=None):
        session = self.alarm.current_session
        try:
            if not session:
                LOGGER.debug("Not authenticating anymore for session id is %s", session_id,)
                return ('', 200)
            elif session.id != session_id:
                LOGGER.debug("Got invalid session id %s, current session id is %s",
                             session_id, session.id)
            elif session.is_authenticated:
                self.alarm.set_disarm_time(0)
            elif not session.is_authenticated:
                events.on_authentication_failed(self.twilio_agent)
        except Exception:
            LOGGER.exception("Failed procession callback for session %s", session_id)

        return ('', 200)

    def auth_request(self, session_id=None):

        LOGGER.debug("Inside authenticated request %s", str(request.form))
        session = self.alarm.current_session
        LOGGER.debug("Session is %s", str(session))
        digits = request.form.get("Digits")
        LOGGER.debug("Got digits %s", str(digits))
        voice_response = VoiceResponse()

        if not session and self.alarm.state != AlarmState.ALARMING:
            # TODO if disarmed get enable time
            voice_response.say(
                "You already have been authenticated. current state is {}".format(self.alarm.state))
            voice_response.say("Goodbye")
            voice_response.hangup()

        elif not session:
            voice_response.say("Alarm is blaring. Goodbye.")
            voice_response.hangup()
            return str(voice_response)

        elif session.id != session_id:
            LOGGER.error("No session found for session id %s", session_id)
            voice_response = GENERIC_ERROR

        elif not session.is_authenticated and not digits:
            voice_response.say("Hi this is your alarm speaking.")
            voice_response.gather(self.get_gather(session.id)).say(
                "Please enter your password followed by the pound key.")

        elif not session.is_authenticated and digits:
            if session.authenticate(self.twilio_agent, digits):
                voice_response.say("You have been authenticated")
                voice_response.gather(self.get_gather(session.id)).say(
                    "Enter disarm time, last digit is the unit followed by the pound key. Or, hang-up now to disable the alarm.")
            else:
                if self.alarm.state != AlarmState.ALARMING:
                    voice_response.say("Authentication failed")
                    voice_response.say(
                        "You have {} tries remaining".format(session.remaining_tries))
                    voice_response.gather(self.get_gather(session.id)).say(
                        "Enter your password followed by the pound key")
                else:
                    voice_response.say("Authentication failed")
                    voice_response.say("Goodbye")
                    voice_response.hangup()

        elif session.is_authenticated and (not digits or len(digits) == 1):
            voice_response.gather(self.get_gather(session.id)).say(
                "Enter disarm time, last digit is the unit followed by the pound key. Or, hang-up now to disable the alarm.")

        elif session.is_authenticated and digits:
            try:
                unit = DIGITS_TO_UNIT[digits[-1]]
                if not unit:
                    voice_response.gather(self.get_gather(session.id)).say(
                        "Enter disarm time, last digit is the unit followed by the pound key. Or, hang-up now to disable the alarm.")
                else:
                    self.alarm.set_disarm_time(digits[:-1]+unit)
                    voice_response.say("Disarm time set for "+self.alarm.get_readable_disarm_time())
                    voice_response.hangup()

            except Exception:
                LOGGER.exception("Failed setting the disarm time")
                voice_response.gather(self.get_gather(session.id)).say(
                    "Enter disarm time, last digit is the unit followed by the pound key. Or, hang-up now to disable the alarm.")
        else:
            voice_response = GENERIC_ERROR

        return str(voice_response)


class Twilio(object):

    def __init__(self, alarm, web_server, auth_delay="0s", account_sid=None, auth_token=None, landline_phone_number=None, mobile_phone_number=None):
        self.auth_delay = parse_duration(auth_delay).total_seconds()
        self.landline_phone_number = landline_phone_number
        self.mobile_phone_number = mobile_phone_number
        self.web_server = web_server
        self.alarm = alarm
        self.client = Client(account_sid, auth_token)
        self.auth_token = auth_token

        self.twilio_server = TwilioServer(self)

        self._register_events_handlers()

        events.authenticator_started(self)

    def _register_events_handlers(self):
        events.alarm_authenticating += self.on_authentication_required

    def on_authentication_required(self, _, session):
        timer = Timer(self.auth_delay, self.trigger_authentication_call, args=[session.id])
        timer.start()

    def trigger_authentication_call(self, session_id):
        # Might check the session_id
        if not self.alarm.current_session:
            LOGGER.debug("Already authenticated, not making the call")
            return
        try:
            self.client.calls.create(
                to=self.mobile_phone_number,
                from_=self.landline_phone_number,
                url=self.twilio_server.get_auth_action_url(session_id),
                status_callback=self.twilio_server.get_status_callback_url(session_id),
                status_callback_event=["completed"]
            )
        except Exception:
            LOGGER.exception("Failed making call for authentication")
            events.authentication_failed(self, session_id, AuthFailureReason.AUTHENTICATOR_FAILURE)
